### PR TITLE
EJBCLIENT-121: Added code to handle concurrent reconnection attempts. Added test case for testing re-connect under concurrent load.

### PR DESCRIPTION
**Pull request by wolfc**
https://issues.jboss.org/browse/EJBCLIENT-121

### CODE REVIEW

**Comment by wolfc** [comment]
Note that I have completely replaced reconnect handling with the code written by @rachmatowicz.

File: src/main/java/org/jboss/ejb/client/EJBClientContext.java [hunk 1]
```diff
@@ -1148,51 +1155,76 @@ private void waitForClusterTopology(final String clusterName) {
 
     private void attemptReconnections() {
         final CountDownLatch reconnectTasksCompletionNotifierLatch;
-        synchronized(this) {
+        final List<ReconnectHandler> reconnectHandlersToAttempt;
+
+        // check for circumstances where we don't want to attempt reconnection
+        synchronized (this) {
```
Comment (dmlloyd) [inline src/main/java/org/jboss/ejb/client/EJBClientContext.java]: It looks like we're combining synchronization, Locks, and atomic/CAS.  Is this intended, or is there a simpler reduction of this code possible?

Comment (rachmatowicz) [inline src/main/java/org/jboss/ejb/client/EJBClientContext.java]: It was intentional, and i'm not sure of there is a simpler way of arranging it.
When a reconnection is required and there are many incoming invocations, all will be redirected through the "attempt reconnections" code path and each will try to initiate a reconnection attempt. I wanted to make one thread (the first one arriving) not only manage the reconnection attempts, but also cause the other threads to block and then control their release after the reconnection completes, in such a way that they do not again try to attempt re-connections (as it has just been completed). Using the monitor lock didn't seem to allow me the possibility of changing the code path of these latter threads, which is why I opted for a separate AtomicBoolean and ReentrantLock/Condition.

**Comment by dmlloyd** [comment]
This PR can no longer target the master branch; please target 2.x instead.

**Pull request closed without merging**